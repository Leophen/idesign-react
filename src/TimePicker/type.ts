export interface TimePickerProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 时间值
   */
  value?: string;
  /**
   * 默认时间值
   * @default 当前时间
   */
  defaultValue?: string;
  /**
   * 触发方式
   * @default click
   */
  trigger?: 'hover' | 'click' | 'context-menu';
  /**
   * 是否禁用选择器
   * @default false
   */
  disabled?: boolean;
  /**
   * 用于格式化时间，[详细文档](https://day.js.org/docs/en/display/format)
   * @default HH:mm:ss
   */
  format?: string;
  /**
   * [时, 分, 秒] 时间间隔步数
   * @default () => [1, 1, 1]
   */
  steps?: Array<string | number>;
  /**
   * 选中时间变化时触发
   */
  onChange?: (value: string) => void;
  /**
   * 切换时间面板时触发
   */
  onTrigger?: (visible: boolean) => void;
}

export interface TimeInputProps {
  /**
   * 时间值
   */
  value?: TimesType;
  /**
   * 是否禁用输入框
   * @default false
   */
  disabled?: boolean;
  /**
   * 用于格式化时间，[详细文档](https://day.js.org/docs/en/display/format)
   * @default HH:mm:ss
   */
  format?: string;
  /**
   * 输入时间变化时触发
   */
  onChange?: (value: TimesType) => void;
}

export interface TimePanelProps {
  /**
   * 是否禁用组件
   * @default false
   */
  disabled?: boolean;
  /**
   * 占位符
   * @default ''
   */
  placeholder?: string;
  /**
   * 尺寸
   * @default medium
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * 选中值
   */
  value?: TimesType;
  /**
   * 时间间隔步数，数组排列 [小时, 分钟, 秒]，示例：[2, 1, 1] 或者 ['2', '1', '1']
   * @default () => [1, 1, 1]
   */
  steps?: Array<string | number>;
  /**
   * 禁用时间项
   */
  disableTime?: (h: number, m: number, s: number) => boolean;
  /**
   * 用于格式化时间，[详细文档](https://day.js.org/docs/en/display/format)
   * @default 'HH:mm:ss'
   */
  format?: string;
  /**
   * 是否隐藏禁用状态的时间项
   * @default true
   */
  hideDisabledTime?: boolean;
  /**
   * 点击此刻按钮时触发
   */
  onNow?: () => void;
  /**
   * 选中值发生变化时触发
   */
  onChange?: (type: string, value: string) => void;
  /**
   * 面板关闭时触发
   */
  onClose?: (visible: boolean) => void;
  /**
   * 点击确认按钮时触发
   */
  onConfirm?: () => void;
}

export interface TimesType {
  hour?: string;
  minute?: string;
  second?: string;
  meridiem?: string;
}

export enum EPickerCols {
  hour = 'hour',
  minute = 'minute',
  second = 'second',
  meridiem = 'meridiem',
}
