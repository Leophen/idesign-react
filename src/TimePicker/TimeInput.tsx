import React, { useState, useEffect } from 'react';
import classNames from 'classnames';
import './index.scss';
import _ from 'lodash';
import Input from '../Input';
import { TimeInputProps } from './type';
import { DEFAULT_FORMAT } from './index'

const TimeInput: React.FC<TimeInputProps> = (props) => {
  const {
    value = { hour: '00', minute: '00', second: '00', meridiem: 'AM' },
    disabled = false,
    format = DEFAULT_FORMAT,
    onChange = () => { }
  } = props

  const [innerValue, setInnerValue] = useState(value)
  const [show, setShow] = useState({
    hour: false,
    minute: false,
    second: false
  })

  useEffect(() => {
    /H|h/.test(format) && (show.hour = true);
    /m/.test(format) && (show.minute = true);
    /s/.test(format) && (show.second = true);
    setShow({ ...show })

    let newHour = /^[H]/.test(format) ? value.hour : Number(value.hour) % 12
    newHour = _.padStart(String(newHour), 2, '0');
    value.hour = newHour
    setInnerValue({ ...value })
  }, [value, format])

  const inputChangeHour = (val: string) => {
    val = _.padStart(String(val), 2, '0');
    innerValue.hour = val
    setInnerValue({ ...innerValue })
    onChange?.(innerValue)
  }

  const inputChangeMinute = (val: string) => {
    val = _.padStart(String(val), 2, '0');
    innerValue.minute = val
    setInnerValue({ ...innerValue })
    onChange?.(innerValue)
  }

  const inputChangeSecond = (val: string) => {
    val = _.padStart(String(val), 2, '0');
    innerValue.second = val
    setInnerValue({ ...innerValue })
    onChange?.(innerValue)
  }

  const handleUsualBlur = () => {
    innerValue.hour = _.padStart(String(innerValue.hour), 2, '0')
    innerValue.minute = _.padStart(String(innerValue.minute), 2, '0')
    innerValue.second = _.padStart(String(innerValue.second), 2, '0')
    setInnerValue({ ...value })
  }

  const timeColon = (
    <div
      className={classNames(
        'i-time-colon',
        disabled && 'i-time-colon__disabled'
      )}
    >
      :
    </div>
  )

  return (
    <Input
      placeholder=""
      readonly={!disabled}
      disabled={disabled}
      suffixIcon="Clock"
    >
      {show.hour && <>
        <Input
          size="small"
          type="number"
          hideNumberBtn
          minNumber={0}
          maxNumber={/^[H]/.test(format) ? 23 : 11}
          placeholder=''
          disabled={disabled}
          value={innerValue?.hour || ''}
          onChange={inputChangeHour}
          onBlur={handleUsualBlur}
        />
        {timeColon}
      </>}
      {show.minute && <>
        <Input
          size="small"
          type="number"
          hideNumberBtn
          minNumber={0}
          maxNumber={59}
          placeholder=''
          disabled={disabled}
          value={innerValue?.minute || ''}
          onChange={inputChangeMinute}
          onBlur={handleUsualBlur}
        />
      </>}
      {show.second && <>
        {timeColon}
        <Input
          size="small"
          type="number"
          hideNumberBtn
          minNumber={0}
          maxNumber={59}
          placeholder=''
          disabled={disabled}
          value={innerValue?.second || ''}
          onChange={inputChangeSecond}
          onBlur={handleUsualBlur}
        />
      </>}
    </Input>
  )
}

export default TimeInput;
