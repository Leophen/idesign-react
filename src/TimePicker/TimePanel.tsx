import React, { useState, useEffect, useCallback, useMemo, useRef } from 'react';
import classNames from 'classnames';
import './index.scss';
import _ from 'lodash';
import dayjs from 'dayjs';
import Button from '../Button';
import { DEFAULT_FORMAT, DEFAULT_STEPS, MERIDIEM_LIST, timeArr } from './index'
import { EPickerCols, TimePanelProps } from './type';

const TimePanel: React.FC<TimePanelProps> = (props) => {
  const {
    value = { hour: '00', minute: '00', second: '00', meridiem: '' },
    format = DEFAULT_FORMAT,
    steps = DEFAULT_STEPS,
    hideDisabledTime = true,
    disableTime,
    onNow = () => { },
    onChange = () => { },
    onClose = () => { },
    onConfirm = () => { }
  } = props

  const [cols, setCols] = useState<Array<EPickerCols>>([]);

  // 初始化 -> 渲染时间数据
  useEffect(() => {
    const match = format.match(/(a\s+|A\s+)?(h+|H+)?:?(m+)?:?(s+)?(\s+a|\s+A)?/);
    const [, startCol, hourCol, minuteCol, secondCol, endCol] = match as any;
    const { meridiem, hour, minute, second } = EPickerCols;
    const renderCol = [
      startCol && meridiem,
      hourCol && hour,
      minuteCol && minute,
      secondCol && second,
      endCol && meridiem,
    ].filter((v) => !!v);
    setCols(renderCol);
  }, [format]);
  const dayjsValue = useMemo(() => {
    const isStepsSet = !!steps.filter((v) => v > 1).length;
    if (value) {
      return dayjs(`${value.hour}-${value.minute}-${value.second}`, format);
    }
    if (isStepsSet) {
      return dayjs()
        .hour(Number(steps[0]) - 1)
        .minute(Number(steps[1]) - 1)
        .second(Number(steps[2]) - 1);
    }
    return dayjs();
  }, [value.hour, value.minute, value.second, format, steps]);
  const getColList = useCallback(
    (col: EPickerCols) => {
      let count = 0;
      if (timeArr.includes(col)) {
        // hour/minute/second column scroll render
        const colIdx = timeArr.indexOf(col);
        const colStep = steps[colIdx];
        if (col === EPickerCols.hour) {
          count = /[h]{1}/.test(format) ? 11 : 23;
        } else {
          count = 59;
        }
        const colList = _.range(
          0,
          count + 1,
          Number(colStep)).map((v) => _.padStart(String(v), 2, '0')
          ) || [];
        return hideDisabledTime && !!disableTime
          ? colList.filter((t) => {
            const params: [number, number, number] = [dayjsValue.hour(), dayjsValue.minute(), dayjsValue.second()];
            params[colIdx] = Number(t);
            return !disableTime?.(...params);
          })
          : colList;
      }
      // meridiem column scroll render
      return MERIDIEM_LIST;
    },
    [steps, format, hideDisabledTime, dayjsValue, disableTime],
  );

  // 同步滚动
  const hourPanelRef = useRef<any>(null)
  const minutePanelRef = useRef<any>(null)
  const secondPanelRef = useRef<any>(null)
  const meridiemPanelRef = useRef<any>(null)
  const getRef = (type: string) => {
    let ref = meridiemPanelRef
    type === 'hour' && (ref = hourPanelRef);
    type === 'minute' && (ref = minutePanelRef);
    type === 'second' && (ref = secondPanelRef);
    return ref
  }
  const updateScroll = (mode?: string) => {
    hourPanelRef.current && hourPanelRef.current.scrollTo({
      top: 32 * (Number(value.hour) / Number(steps[0])),
      behavior: mode
    });
    minutePanelRef.current && minutePanelRef.current.scrollTo({
      top: 32 * (Number(value.minute) / Number(steps[1])),
      behavior: mode
    });
    secondPanelRef.current && secondPanelRef.current.scrollTo({
      top: 32 * (Number(value.second) / Number(steps[2])),
      behavior: mode
    });
  }
  useEffect(() => {
    setTimeout(() => {
      updateScroll()
    })
  }, []);
  useEffect(() => {
    updateScroll('smooth')
  }, [value]);

  const clickItem = (type: string, val: string) => {
    onChange?.(type, val)
  }

  const handleNow = () => {
    onNow?.()
  }

  const closePanel = () => {
    onClose?.(false)
  }

  const handleConfirm = () => {
    closePanel()
    onConfirm?.()
  }

  return (
    <div
      className="i-time-panel"
    >
      <section
        className="i-time-panel-content"
      >
        {cols.map((col, index) => (
          <ul
            className="i-time-panel-item"
            key={col + index}
            ref={getRef(col)}
          >
            {getColList(col).map((item) => (
              <li
                className={classNames(
                  'i-time-spinner-item',
                  (value as any)[col] === item && 'i-time-spinner-item__selected'
                )}
                key={item}
                onClick={() => clickItem(col, item)}
              >
                {item}
              </li>
            ))}
          </ul>
        ))}
      </section>
      <footer className="i-time-panel-footer">
        {!steps.filter((v) => v > 1).length && <Button size="small" variant="text" onClick={handleNow}>此刻</Button>}
        <div className="i-time-panel-footer__handle">
          <Button size="small" variant="outline" onClick={closePanel}>取消</Button>
          <Button size="small" onClick={handleConfirm}>确认</Button>
        </div>
      </footer>
    </div>
  )
}

export default TimePanel;
