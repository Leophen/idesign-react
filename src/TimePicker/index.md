---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 7
---

# TimePicker 时间选择器

用于选择指定时间。

## 基本用法

```tsx
import React, { useState } from 'react'
import { TimePicker } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('12:34:56')

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <h4>无默认值</h4>
      <TimePicker />
      <h4>有默认值</h4>
      <TimePicker defaultValue={value} />
      <h4>有受控值</h4>
      <TimePicker value={value} />
      <h4>一般用法</h4>
      <TimePicker
        value={value}
        onChange={handleChange}
      />
    </>
  )
}

export default App
```

## 禁用状态

可通过 `disabled` 属性控制为禁用状态。

```tsx
import React from 'react'
import { TimePicker } from 'idesign-react'

const App = () => {
  return (
    <TimePicker disabled />
  )
}

export default App
```

## 不同触发方式

可通过 `trigger` 属性指定下拉菜单触发方式，默认为 `click`。

```tsx
import React from 'react'
import { TimePicker } from 'idesign-react'

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <TimePicker trigger="click" />
      <TimePicker trigger="hover" />
      <TimePicker trigger="context-menu" />
    </div>
  )
}

export default App
```

## 不同时间类型

可通过 `format` 属性指定不同的时间类型，默认为 `HH:mm:ss`，[点击查看各类型详情](https://day.js.org/docs/en/display/format)。

```tsx
import React, { useState } from 'react'
import { TimePicker } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('12:34:56')

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <h4>24小时制 - 时分秒选择器</h4>
      <TimePicker format="HH:mm:ss" value={value} onChange={handleChange} />
      <h4>24小时制 - 时分选择器</h4>
      <TimePicker format="HH:mm" value={value} onChange={handleChange} />
      <h4>12小时制 - 时分秒选择器</h4>
      <TimePicker format="hh:mm:ss A" value={value} onChange={handleChange} />
      <h4>12小时制 - 时分选择器</h4>
      <TimePicker format="hh:mm" value={value} onChange={handleChange} />
      <h4>分秒选择器</h4>
      <TimePicker format="mm:ss" value={value} onChange={handleChange} />
    </>
  )
}

export default App
```

## 自定义步长

可通过 `steps` 属性指定时间面板选择步长，默认为 `[1, 1, 1]`。

```tsx
import React from 'react'
import { TimePicker } from 'idesign-react'

const App = () => {
  return (
    <>
      <h4>默认步长</h4>
      <TimePicker steps={[1, 1, 1]} />
      <h4>分秒步长设为 5</h4>
      <TimePicker steps={[1, 5, 5]} />
    </>
  )
}

export default App
```

<API />
