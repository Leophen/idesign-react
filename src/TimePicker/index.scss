.i-time-popup {
  section > .i-input {
    & > .i-input {
      width: 20px;
      padding: 0;
      border: none;
      .i-input__inner {
        padding-right: 0;
        text-align: center;
      }
    }
    & > .i-input__inner,
    .i-input-number-slider {
      display: none;
    }
    & > .i-input-suffix-icon {
      pointer-events: none;
    }
  }
}

.i-time-colon {
  margin-bottom: 2px;
  color: var(--i-font-2);
  &.i-time-colon__disabled {
    color: var(--i-font-disabled);
  }
}

.i-time-panel {
  width: 200px;
}

.i-time-panel-content {
  display: flex;
  align-items: center;
  height: 200px;
}

.i-time-panel-item {
  flex: 1;
  height: 100%;
  margin: 0;
  padding: 0;
  overflow: auto;
  overflow: overlay;
  &::-webkit-scrollbar {
    display: none;
  }
  &::after {
    display: list-item;
    width: 100%;
    height: 168px;
    content: '';
  }
}

.i-time-spinner-item {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 32px;
  font-size: 12px;
  line-height: 32px;
  list-style: none;
  cursor: default;
  &:hover {
    background: var(--i-bg-hover);
  }
  &.i-time-spinner-item__selected {
    font-weight: bold;
    background: var(--i-bg-active);
    &:hover {
      background: var(--i-bg-active);
    }
  }
}

.i-time-panel-footer {
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  width: 100%;
  padding: 12px;
  border-top: 1px solid var(--i-border);
}

.i-time-panel-footer__handle {
  display: flex;
  gap: 8px;
  align-items: center;
  margin-left: auto;
}
