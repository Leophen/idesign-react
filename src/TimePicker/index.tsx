import React, { useState, useEffect } from 'react';
import classNames from 'classnames';
import './index.scss';
import dayjs from 'dayjs';
import Popup from '../Popup';
import useDefault from '../hooks/useDefault';
import TimePanel from './TimePanel';
import TimeInput from './TimeInput';
import { EPickerCols, TimePickerProps, TimesType } from './type';

export const MERIDIEM_LIST = ['AM', 'PM'];
export const DEFAULT_STEPS = [1, 1, 1];
export const DEFAULT_FORMAT = 'HH:mm:ss';
export const timeArr = [EPickerCols.hour, EPickerCols.minute, EPickerCols.second];

const TimePicker: React.FC<TimePickerProps> = (props) => {
  const {
    className,
    style,
    value,
    defaultValue = dayjs().format(props.format || DEFAULT_FORMAT),
    trigger = "click",
    disabled = false,
    format = DEFAULT_FORMAT,
    steps = DEFAULT_STEPS,
    onChange = () => { },
    onTrigger = () => { }
  } = props;

  const getCurrentTime = (type?: string) => {
    let currentVal = new Date().getHours().toString()
    if (type === 'meridiem') {
      /H|h/.test(format) && (currentVal = innerValue.split(':')[0])
      Number(currentVal) >= 12 ? (currentVal = 'PM') : (currentVal = 'AM')
    } else {
      type === 'minute' && (currentVal = new Date().getMinutes().toString());
      type === 'second' && (currentVal = new Date().getSeconds().toString());
      (currentVal.length === 1) && (currentVal = '0' + currentVal);
    }
    return currentVal
  }

  const [innerValue, setInnerValue] = useDefault(value, defaultValue, onChange);

  const [timeValue, setTimeValue] = useState<any>({
    hour: '00',
    minute: '00',
    second: '00',
    meridiem: ''
  })

  const valueToObj = (val: string) => {
    const result = {
      hour: '00',
      minute: '00',
      second: '00',
      meridiem: ''
    }
    let ifMeridiem = false;
    /A/.test(format) && (ifMeridiem = true);
    const timeArr = val.split(':')
    if (!/H|h/.test(format) && timeArr.length === 2) {
      timeArr.unshift('00')
    }
    result.hour = timeArr[0];
    result.minute = timeArr[1];
    result.second = String(timeArr[2]).split(' ')[0];
    ifMeridiem && (result.meridiem = val.split(' ')[1] || getCurrentTime('meridiem'))
    return result
  }

  const objToValue = (obj: TimesType) => {
    let hourVal = `${obj.hour}:`
    let minuteVal = `${obj.minute}`
    let secondVal = `:${obj.second}`
    let meridiemVal = ``

    !/H|h/.test(format) && (hourVal = ``);
    !/s/.test(format) && (secondVal = ``);
    /A/.test(format) && (meridiemVal = ` ${timeValue.meridiem}`);

    const result = `${hourVal}${minuteVal}${secondVal}${meridiemVal}`
    return result
  }

  useEffect(() => {
    const newTimeVal = valueToObj(innerValue)
    setTimeValue({ ...newTimeVal })
  }, [innerValue])

  const updateValue = (val?: TimesType) => {
    let currentTime
    val ? (currentTime = objToValue(val)) : (currentTime = objToValue(timeValue))
    setInnerValue(currentTime)
  }

  const selectTime = (type: string, val: string) => {
    timeValue[type] = val
    updateValue()
  }

  const [popupVisible, setPopupVisible] = useState(false)
  const switchPopup = (visible: boolean) => {
    setPopupVisible(visible)
    onTrigger?.(visible)
  }

  const handleNow = () => {
    const currentTime = dayjs().format(format)
    setInnerValue(currentTime)
    onChange?.(currentTime)
  }

  const handleConfirm = () => {
    switchPopup(false)
    updateValue()
  }

  return (
    <div
      className={classNames(
        'i-time-picker',
        className
      )}
      style={{ ...style }}
    >
      <Popup
        content={
          <TimePanel
            value={timeValue}
            format={format}
            steps={steps}
            onNow={handleNow}
            onClose={switchPopup}
            onChange={selectTime}
            onConfirm={handleConfirm}
          />
        }
        placement="bottom"
        trigger={trigger}
        className="i-time-popup"
        visible={popupVisible}
        disabled={disabled}
        onTrigger={switchPopup}
      >
        <TimeInput
          value={timeValue}
          format={format}
          disabled={disabled}
          onChange={updateValue}
        />
      </Popup>
    </div>
  );
};

TimePicker.displayName = 'TimePicker';

export default TimePicker;
