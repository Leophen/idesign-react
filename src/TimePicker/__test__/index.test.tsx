import { render } from '@testing-library/react';
import TimePicker from '../index';

describe('TimePicker 组件测试', () => {
  const renderTimeVal = (comp) => {
    const { container } = render(comp);
    const inputList = container.querySelectorAll('.i-input__inner')
    return `${inputList[0].value}:${inputList[1].value}:${inputList[2].value}`
  }

  it('create', () => {
    const { asFragment } = render(<TimePicker value='12:34:56' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('value', () => {
    expect(renderTimeVal(<TimePicker value='12:34:56' />)).toEqual('12:34:56');
  });

  it('defaultValue', () => {
    expect(renderTimeVal(<TimePicker defaultValue='12:34:56' />)).toEqual('12:34:56');
  });

  it('disabled', () => {
    const { container } = render(<TimePicker disabled />);
    expect(container.querySelector('.i-input')).toHaveClass('i-input-is-disabled');
  });
});

