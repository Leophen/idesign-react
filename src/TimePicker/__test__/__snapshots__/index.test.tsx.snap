// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-time-picker"
  >
    <div
      class="i-popup__reference i-time-popup"
    >
      <div
        class="i-input i-input-is-readonly"
      >
        <div
          class="i-input i-input--size-small i-input--type-number"
        >
          <input
            class="i-input__inner"
            max="23"
            min="0"
            placeholder=""
            step="1"
            type="number"
            value="12"
          />
          <div
            class="i-input-number-slider"
          />
        </div>
        <div
          class="i-time-colon"
        >
          :
        </div>
        <div
          class="i-input i-input--size-small i-input--type-number"
        >
          <input
            class="i-input__inner"
            max="59"
            min="0"
            placeholder=""
            step="1"
            type="number"
            value="34"
          />
          <div
            class="i-input-number-slider"
          />
        </div>
        <div
          class="i-time-colon"
        >
          :
        </div>
        <div
          class="i-input i-input--size-small i-input--type-number"
        >
          <input
            class="i-input__inner"
            max="59"
            min="0"
            placeholder=""
            step="1"
            type="number"
            value="56"
          />
          <div
            class="i-input-number-slider"
          />
        </div>
        <input
          class="i-input__inner"
          placeholder=""
          readonly=""
          value=""
        />
        <i
          class="i-icon icon-Clock i-input-suffix-icon"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
